QueryPlan Optimizer aims to run TPCH 100x faster

auxten opened the pull request:
Still in a very early stage. This PR is only for tracking progress.

auxten committed:
Gen manifest

auxten committed:
Upload pypi only on v tag

auxten committed:
Make mac-arm64 for apple silicon(M1, M2)

nmreadelf committed:
Added Dataframe output format (#6)

auxten committed:
Add to pandas example (#5)

auxten committed:
Add arch chart

auxten committed:
Fix readme

lmangani committed:
AWS Lambda compatibility (#9)

* PR_SET_NAME workaround

AWS Lambdas (and other virtualized platforms) lack of support for PR_SET_NAME causing a blocking exception. Pending an upstream PR or fix in ClickHouse, this patch allows this function to fail unharmed. The resulting executable has been tested on various platforms without drawbacks and discussed in clickhouse issue [29378](https://github.com/ClickHouse/ClickHouse/issues/29378)

$ sed -i '/Cannot set thread name/c\' /ClickHouse/src/Common/setThreadName.cpp

* Disable AVX2 support

auxten committed:
Fix build_mac_arm64.sh (#10)

auxten committed:
Fix version in __init__.py

auxten committed:
Fix replace file content bug

auxten committed:
Try import pyarrow and pandas when necessary

auxten committed:
Build on pr based on pybind and push tags

auxten committed:
Sleep 60 then kill clang if timeout

auxten committed:
Update brew to avoid llvm@15 minor version unstable

auxten committed:
Add serverless demo and discord

lmangani committed:
Enable S3

Proposal to enable S3 features to improve remote file reading performance compared to plain url engine.

auxten committed:
Add benchmark in readme

auxten committed:
Brew update before install
Although we lock llvm@15, different github action runner may have llvm 15.0.7 or 15.0.7_1.
This will cause ccache miss a lot. Hope brew update will keep llvm latest

auxten committed:
Brew update before install
Although we lock llvm@15, different github action runner may have llvm 15.0.7 or 15.0.7_1.
This will cause ccache miss a lot. Hope brew update will keep llvm latest

auxten committed:
Remove timestamp for ccache files

auxten committed:
Fix ArrowTable output via '-m chdb sql arrowtable'

auxten committed:
Fix up tests

auxten committed:
Update build_wheels.yml

auxten committed:
Update build_wheels.yml

auxten committed:
Test 60 types of output and run tests in CI

auxten committed:
Run tests on linux build

auxten committed:
Fix dup run tests

auxten committed:
Use assertAlmostEqual for test output

auxten committed:
Return non zero if any test fails

auxten committed:
Test ArrowTable, because Arrow binary is not stable cross plat

auxten committed:
Ignore tz issue in test

auxten committed:
Install pyarrow pandas before make test

auxten committed:
Fix CIBW_TEST

auxten committed:
Add branch name in ccache key to keep default branch a cache
See: https://github.com/actions/cache/blob/main/tips-and-workarounds.md#use-cache-across-feature-branches

auxten committed:
Trigger action when push to pybind

auxten committed:
Remove ref name in cache

auxten committed:
Test enable jemalloc in linux x86_64 build

auxten committed:
Fix parallel test and test more in mac arm64 build (#21)

lmangani committed:
Add function: chdb() to return version (#25)

* add function: chdb() to return version

This is an example of a C++ function that returns the current version of chDB as a string. The version number is defined using the CHDB_VERSION_STRING macro, which should be defined during the build process.

The FunctionChdbVersion class is a subclass of FunctionConstantBase, which is a base class for functions that return a constant value. The FunctionChdbVersion constructor takes the CHDB_VERSION_STRING macro as an argument, which is passed to the FunctionConstantBase constructor to set the constant value of the function.

The REGISTER_FUNCTION macro is used to register the FunctionChdbVersion function with the FunctionFactory. The REGISTER_FUNCTION macro should be conditionally compiled using an #if directive that checks whether the CHDB_VERSION_STRING macro is defined, to avoid registering the function if the version number is not available.

* Update chdbVersion.cpp

* Update chdbVersion.cpp

* Update CMakeLists.txt

* Update CMakeLists.txt

lmangani committed:
Add server example (#27)

Basic example using chdb to serve GET/POST API requests, compatible with ClickHouse Play client queries. 

No action needed as this is not part of the compiled code.

Commit:
Add format example in readme code (#24)

* Add format example in readme code

* change readme.md, use Pretty for simple example

* use utf-8 decode for json print in readme.md

---------

Co-authored-by: laodouya <xujin3@xunlei.com>

auxten committed:
Chinese readme (#29)

* Add chinese readme

* Add link

auxten committed:
Try macos-12-xl

auxten committed:
Use auxten/jemalloc for mallctl("arenas.lookup", NULL, NULL, &ptr, sizeof(ptr))

auxten committed:
Add mallocAdapt with mem alloc wrapper and libc mem apis

auxten committed:
Add __real_free stub to make clickhouse-local compile succ

auxten committed:
Move global inside_main to LocalChdb.cpp

auxten committed:
Use je_xx funcs when USE_JEMALLOC defined

auxten committed:
Remove jemalloc_rename.h

auxten committed:
Use je_ prefix for jemalloc

auxten committed:
Inject jemalloc symbols for necessary libs

auxten committed:
Use ordinary macos-12

auxten committed:
Fix build.sh for linux to enable jemalloc

auxten committed:
Fix #if defined(USE_JEMALLOC) && (USE_JEMALLOC)

auxten committed:
Fix typo

auxten committed:
Update auxten/jemalloc fix if ptr present check alloc_ctx.edata == NULL

auxten committed:
Add timestamp for ccache to fix "another job may be creating this cache"

auxten committed:
Fix output buffer flush caused memory realloc x2

auxten committed:
Add test for issue #31

auxten committed:
Add test for issue #31

auxten committed:
Limit output of test

auxten committed:
Make select result stable

Commit:
Add db api 2.0 driver

auxten committed:
Fix dbapi feature packaging

auxten committed:
Fix minor bug

Commit:
Remove arrow in dbapi

auxten committed:
Readme

lmangani committed:
Action: ignore readme, examples, workflows

auxten committed:
Add ArrowStream Arrow format

auxten committed:
Add Table and run chdb on Table

auxten committed:
Test to_pandas()

auxten committed:
Add hits_0.parquet download

auxten committed:
Arrow format output is not deterministic, check the first 100 bytes

auxten committed:
Ok, 10 bytes

auxten committed:
Not check arrow

auxten committed:
Drop Python 3.7 support

auxten committed:
Use total downloads

auxten committed:
Warn if pandas version lower then 2.0.0

auxten committed:
Use pandas_read_parquet

auxten committed:
Add flush_to_disk for Table

auxten committed:
Fix fd leak in tests

auxten committed:
Update usage in readme

auxten committed:
Fix typo

lmangani committed:
Improve format of data options

lmangani committed:
Improve data input style README-zh.md

auxten committed:
Update README.md

auxten committed:
Update README-zh.md

auxten committed:
Use apache 2.0 license

auxten committed:
Let pandas choose backend itself

auxten committed:
Check memoryview size before using

auxten committed:
Check format is supported before the engine runs too far

lmangani committed:
Restart Failed Actions

lmangani committed:
Require secret.PAT_TOKEN

@auxten we need a `PAT_TOKEN` secret for this action please

lmangani committed:
Update build_wheels.yml

lmangani committed:
Update build_wheels.yml

lmangani committed:
Update build_wheels.yml

auxten committed:
Fix logo blur

lmangani committed:
Add Colab Examples

auxten committed:
Fix nullptr caused core dump

lmangani committed:
Fix discord link

auxten committed:
Change default branch to main

auxten committed:
Check self.use_memfd

auxten committed:
Fix missing endif()

auxten committed:
Update format output for parquet Bool support

auxten committed:
Add contributors section

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: create .all-contributorsrc [skip ci]

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

auxten committed:
Update README.md

lmangani committed:
Format README.md

auxten committed:
Update README.md

auxten committed:
Enable brotli

auxten committed:
Add necessary alltypes_dictionary.parquet

auxten committed:
Fix test data path

auxten committed:
Fix data path in example

auxten committed:
Do nothing on StaticThreadPool re-initialize

auxten committed:
Use je_mallctl in zookeeper

auxten committed:
Cleanup unnecessary docs

auxten committed:
Refactor LocalChdb:

  - Fix memoryview gc issue

  - Add __str__ __len__ __repr__ for query_result

  - Add "Debug" as a format to enable ClickHouse debug mode

  - Use `query_result` and `memoryview_wrapper` to hold `local_result_wrapper` with shared_ptr`

  - C++ style `queryToBuffer`

auxten committed:
Use .byte() not .get_memview().tobytes()

auxten committed:
Make free_result not delete result pointer

auxten committed:
Fix tests

auxten committed:
Add pandas dataframe example

auxten committed:
Update readme

auxten committed:
Fix all get_memview.tobytes() to just bytes(), and len(res) directly

lmangani committed:
Do not build on changes to the tests/ folder

auxten committed:
Optimize test_parallel

auxten committed:
Fix tls issue and use PIC and PIE

auxten committed:
Disable thin-lto for now

auxten committed:
Use -flto=thin in release build

auxten committed:
Max sleep time to 120s for easy debug

auxten committed:
Empty printHelp

auxten committed:
Init NamedCollections after setup users

auxten committed:
Add test on issue 60

auxten committed:
Fix test issue 60 url

auxten committed:
Add query_stateful low level func

auxten committed:
Update CONTRIBUTING.md

lmangani committed:
delay before restarting failed actions

reema93jain committed:
Add files via upload

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

auxten committed:
Update README.md

auxten committed:
Add UUID generate_from_pid()

auxten committed:
Add SharedPtrContextHolder(todo: fix leak shared)

auxten committed:
Use pid as Session id

auxten committed:
Add ignore

auxten committed:
Use createSharedHolder to create shared_context in Client

auxten committed:
Free query_result_memory vector only in `free_result`

auxten committed:
Set background_schedule_pool_size 8

auxten committed:
No exception if table uuid exist for stateful query

auxten committed:
Init query_result_memory to nullptr on construct

auxten committed:
Set max_thread_pool_free_size = 0 and use a leaky shared_context for now

auxten committed:
Add chdb-todo: fix SharedPtrContextHolder.shared leak

auxten committed:
Test for gc result on empty output

auxten committed:
Add test_final_join

auxten committed:
Disable BackgroundCleanup

auxten committed:
Set background_schedule_pool_size 0

auxten committed:
Remove stateful tests for now

auxten committed:
Update readme

auxten committed:
New arch chart

auxten committed:
Fix 0.11 issues2 (#67)

* Tests need psutil

* Check thread count only in standalone test run

* Ignore

auxten committed:
Created using Colaboratory

auxten committed:
Created using Colaboratory

auxten committed:
Remove dup

auxten committed:
Update chDB_demos.ipynb

auxten committed:
Fix arrow jemalloc (#68)

* Add bak

* Use patch arrow lib in https://github.com/auxten/arrow

* Enable jemalloc for arrow on Linux

* Define ARROW_JEMALLOC=1

auxten committed:
Enable jemalloc only Linux

auxten committed:
Add memory_pool_jemalloc.cc only Linux

auxten committed:
Fix typo

lmangani committed:
Add Stateful Session Example

lmangani committed:
Update README.md

auxten committed:
Add session example in readme-ch

auxten committed:
Try fix jemalloc on aarch64 (#69)

lmangani committed:
add chDB-GPT Example

auxten committed:
Add some logo

auxten committed:
Only enable AVX and JIT on x86_64

auxten committed:
Add Demo chDB at [ClickHouse v23.7 livehouse!]

auxten committed:
Fix data dir name in ChDB_TPCH_SF10_Queries

lmangani committed:
Broader arm64 compatibility (#77)

auxten committed:
Fix blog link

lmangani committed:
Plain ARM64 builder

lmangani committed:
Delete build_wheels_arm64.yml

nmreadelf committed:
Add arm64 linux build (#70)

* Add arm build

* Update build_arm_whells.yml

* fix build wheel

* Cleanup Job

Cleanup Job for Self-Hosted runners to reclaim disk space after successful or failed builds.

---------

Co-authored-by: Lorenzo Mangani <lorenzo.mangani@gmail.com>

lmangani committed:
Allow module publishing on TAG_NAME versions

lmangani committed:
Update README.md

lmangani committed:
Fix twine upload path

lmangani committed:
Remove test action

lmangani committed:
Populate global UDF settings ClickHouse PR #52643 (#81)

* Backport #52643

Backporting changes from CH #52643

https://github.com/ClickHouse/ClickHouse/pull/52643/files

* Update positional options in Client

Backporting changes from CH #52643

https://github.com/ClickHouse/ClickHouse/pull/52643/files

auxten committed:
Fix stateful mergetree (#83)

auxten committed:
Create FUNDING.yml

auxten committed:
Update FUNDING.yml

lmangani committed:
Update README

Add legal disclaimer

lmangani committed:
Update README.md

auxten committed:
Query on multiple Pandas DataFrame (#89)

* Add dataframe.query() to run sql on multi dataframe

* Add join dataframe test

* Support passing dataframe as table

auxten committed:
Query stats (#91)

* Add query result statistics rows_read(), bytes_read(), elapsed()

* Add example and statistics test

auxten committed:
Fix null

auxten committed:
Add statistics funcs for Table

nmreadelf committed:
Add CHDB_VERION for cmake (#96)

nmreadelf committed:
enable output_format_arrow_string_as_string (#98)

fix #92

auxten committed:
Generate xml and py files for UDF

auxten committed:
Unified chdb.query with udf_path support

auxten committed:
Use query instead of query_stateful

auxten committed:
Use new logo

lmangani committed:
Strip library (#99)

As we grow in arch coverage and versions, space matters. 

Stripping the library could save 100+ megabytes per release/package and the impact on our ability to troubleshoot should be limited since we can easily rebuild locally as needed or bind this to an ENV variable if required.

auxten committed:
Use snake logo

auxten committed:
Use white bg logo

auxten committed:
Add dark theme logo

auxten committed:
Fix dark theme

auxten committed:
All center

lmangani committed:
chdb + urleng session persistence example

auxten committed:
Use output_format as cpp func 2nd arg

auxten committed:
Fix up query default args value

auxten committed:
Use temp dir for UDF config and executables

auxten committed:
Support `with Session() as sess:` in chdb session

auxten committed:
Add tests for UDF and with session

auxten committed:
Add Query with UDF

auxten committed:
Merge pull request #100 from chdb-io/binUdf

Support UDF in Python

lmangani committed:
Create udf.py

lmangani committed:
Update README.md

auxten committed:
Fix case of arrow

auxten committed:
Created using Colaboratory

auxten committed:
No strip, which will casue wired things

auxten committed:
Add diff dir script

auxten committed:
Add ByConity original optimizer stuff

auxten committed:
Adapt optimizer patch no.1

auxten committed:
Fix src/Analyzers/QueryRewriter.cpp

auxten committed:
Diff dir in python

auxten committed:
Better utils/difftree.py

auxten committed:
Merging src/QueryPlan and src/Processors/QueryPlan

auxten committed:
Remove useless proto

auxten committed:
Rename src/QueryPlan

auxten committed:
Ignore tmp bc stuff

auxten committed:
Replace QueryPlan with Processors/QueryPlan

auxten committed:
Update QueryPipeline to QueryPipelineBuilder

auxten committed:
Implement getType, setInputStreams, copy of IQueryPlanStep for QueryPlan steps

auxten committed:
Add elapsedMillisecondsAsDouble

auxten committed:
Adapt optimizer patch no.2:
    1. Add missing transforms
    2. Disable IQueryPlanStep operator==
    3. virtual void updateOutputStream() throw Exception
    4. Remove ASTSerDerHelper

auxten committed:
Adapt optimizer patch no.3:
    1. Throw exception on serialize or deserialize
    2. Fix JoinStep and JoiningTransform
    3. Drop PlanCache
    4. Fix chdbVersion

djouallah commented:
wait are you going to build a new Query optimizer by yourself ?

auxten commented:
> wait are you going to build a new Query optimizer from scratch ?

No, transplant optimizer from ByConity which has forked from ClickHouse v21.8

alanpaulkwan commented:
Is there any way to implement this in the main branch of Clickhouse as well? Would be huge

auxten commented:
> Is there any way to implement this in the main branch of Clickhouse as well? Would be huge

Perhaps later, even though it has already taken me months, I am unsure if this will be successful.😅

djouallah commented:
@auxten you can do it !!!

alexey-milovidov commented:
It makes sense to also send this PR to the main repository - interesting if it will pass the tests. Although I expect it is going to be difficult...

auxten committed:
Ignore join_using_null_safe for now

auxten committed:
Replace all ASOF::Inequality with ASOFJoinInequality

auxten committed:
Add unfinished NestedLoopJoin

auxten committed:
Adapt optimizer patch no.4:
    1. Fix all ASOF::Inequality to ASOFJoinInequality
    2. Remove Exchange and RemoteExchange stuff
    3. Remove DynamicFilter
    4. Remove ProjectionStep

alexey-milovidov commented:
One downside: if it is not tested by ClickHouse CI, it will most likely contain a ton of bugs, and having these bugs will lead to reputational risks for ClickHouse. It means - we have to send a PR to the main ClickHouse repository.

lmangani commented:
> One downside: if it is not tested by ClickHouse CI, it will most likely contain a ton of bugs, and having these bugs will lead to reputational risks for ClickHouse. It means - we have to send a PR to the main ClickHouse repository.

For sure. This will be considered highly experimental and it won't have mainstream ambitions unless we all agree the results are overwhelmingly positive. On the other hand as a fork, we can afford taking some risks the mainline project cannot 😉

auxten commented:
Absolutely agree with @alexey-milovidov. As a data tool the quality is the most important thing. This branch is very very experimental. I have never expected it would draw so much attention.
I'm not even very confident on this could barely run. The quality of chdb depends mostly on ClickHouse. If this experiment could run(god bless), I will try to make a PR to clickhouse main repo. If I can manage it to pass all the ClickHouse tests and reviewed(god bless again). Then we can say it works.
It may take years, but I am happy working on this.

auxten committed:
Adapt optimizer patch no.5:
    1. Remove FinishSortingStep
    2. Fix FilterStep
    3. Fix ProjectionMatchContext
    4. Fix TableScanExecutor::TableScanExecutor
    5. Just keep serialize/deserialize throw Exception

auxten committed:
Cleanup to rebase merge into previous commit

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/chdb-io/chdb?pullRequest=116) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/chdb-io/chdb?pullRequest=116) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: auxten<br/>:x: lmangani<br/>:x: nmreadelf<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/chdb-io/chdb?pullRequest=116) it.</sub>

alexey-milovidov commented:
@auxten Rebase?

auxten commented:
> @auxten Rebase?

I didn't working on this for a long time. It's still very far from general available.
Maybe I should just close this PR
